pkg/kubelet/pod_workers.go

```diff
@@ -281,15 +366,10 @@ type podSyncStatus struct {
 	// it has been killed. When known pods are synced, any pod that is
 	// terminated and has restartRequested will have its history cleared.
 	restartRequested bool
-	// notifyPostTerminating will be closed once the pod transitions to
-	// terminated. After the pod is in terminated state, nothing should be
-	// added to this list.
-	notifyPostTerminating []chan<- struct{}
-	// statusPostTerminating is a list of the status changes associated
-	// with kill pod requests. After the pod is in terminated state, nothing
-	// should be added to this list. The worker will execute the last function
-	// in this list on each termination attempt.
-	statusPostTerminating []PodStatusFunc
+	// observedRuntime is true if the pod has been observed to be present in the
+	// runtime. The pod worker cannot short circuit unstarted pods that have been
+	// observed at runtime.
```

Comment:
Yeah, short circuit is "canEverStart" and the other is "canStart" - anything waiting for "canStart" by definition hasn't been started yet, so the short circuit is killing those right away.  However, if we detect a runtime only pod (RunningPod set) we haven't started that either (status.startedAt is 0, and startPodSync exits early in RunningPod != nil) we can't check startedAt. 

Technically a runningPod has a startedAt, so that may be a different way to represent this.

Reply:
I see that makes sense, and it also makes sense why in `startPodSync` we first check and handle the runtime pod case before we call `allowPodStart`.

What is confusing me here is the comment regarding pod worker **cannot** short circuit part, because it seems that exactly what is happening (i.e. it is short circuiting because for runtime pods we are ignoring `canEverStart` and `canStart`.

Reply:
The short circuit happens in SyncKnownPods - if an orphaned pod hasn't been observed at runtime and hasn't been started, removeTerminatedWorker can simply delete it right away.  If we *have* observed it at runtime, we have to make the attempt to call syncTerminatedRuntimePod at least once.  The new TestUpdatePod has both of these cases covered.

Reply:
I can update the comment.